corfudb-tools/src/main/java/org/corfudb/compactor/CorfuStoreCompactor.java

```diff
@@ -0,0 +1,313 @@
+        return appendCheckpoint(corfuTable,
+                                TableName.newBuilder()
+                                    .setNamespace("")
+                                    .setTableName(tableName)
+                                    .build());
+    }
+
+    private<K, V> Token appendCheckpoint(CorfuTable<K, V> corfuTable, TableName tableName) {
+        MultiCheckpointWriter<CorfuTable> mcw = new MultiCheckpointWriter<>();
+        mcw.addMap(corfuTable);
+
+        long tableCkptStartTime = System.currentTimeMillis();
+        log.info("Starting checkpoint namespace: {}, tableName: {}",
+                tableName.getNamespace(), tableName.getTableName());
+
+        Token trimPoint = mcw.appendCheckpoints(corfuRuntime, "checkpointer");
+
+        long tableCkptEndTime = System.currentTimeMillis();
+        log.info("Completed checkpoint namespace: {}, tableName: {}, with {} entries in {} ms",
+                tableName.getNamespace(),
+                tableName.getTableName(),
+                corfuTable.size(),
+                (tableCkptEndTime - tableCkptStartTime));
+
+        return trimPoint;
+    }
+
+    private Token appendUfoCheckpoint(TableName tableName,
+                                      ISerializer serializer,
+                                      boolean diskBacked) {
+        CorfuTable<CorfuDynamicKey, OpaqueCorfuDynamicRecord> corfuTable = openTable(tableName,
+                                                                               serializer,
+                                                                               diskBacked);
+        return appendCheckpoint(corfuTable, tableName);
+    }
+
+    public void trim() {
+        if(!enableTrim) {
+            return;
+        }
+
+        log.info("enable trim is true, performing prefixTrim");
+        try {
+            trimLog();
+        } catch (WrongEpochException wee) {
+            // TODO: Find a way to log this to syslog
+            log.error(
+                    "Compactor: Trim failed, ignore the WrongEpochException" +
+                            CORFU_LOG_TRIM_ERROR, wee);
+        } catch (Throwable t) {
+            String msg = "Compactor: Trim failed!";
+            // TODO: Find a way to log this to syslog with error code
+            log.error(msg, CORFU_LOG_TRIM_ERROR, t);
+            throw new RuntimeException(msg, t);
+        }
+    }
+
+    // Following stuff mostly duplicates stuff from FrameworkCompactor
+    // refactor to somewhere common and share or merge or something
+
+    /**
+     * Perform log-trimming on CorfuDB by selecting the smallest value from checkpoint map.
+     */
+    private void trimLog() {
+        log.info("Starting CorfuStore trimming task");
+
+        final CorfuTable<String, Token> nodeTrimTokenMap = getNodeTrimTokenMap();
+        final Token thisNodeTrimToken = nodeTrimTokenMap.get(thisNodeUuid);
+
+        if (thisNodeTrimToken == null) {
```

Comment:
can we use Optional instead checking for null?

Reply:
sure, the nodeTrimMap is a corfu table under the hood, but we can wrap the null checks in an Optional